feat: Fix individual song covers and track numbering display

gurnoorpannu opened the pull request:
Closes: #1823

Changes Made
RetroGlideExtension.kt:
Uses individual song artwork (AudioFileCover(song.data)) as first priority.

AudioFileCoverFetcher.kt:
Improved error handling: robust fallback flow, gracefully handles failures so one broken file doesn't affect others.

AudioFileCoverUtils.java:
More comprehensive fallback logic with validation.
Extracts from ID3v2 and ID3v1 tags, checks known external cover files in folder.

MusicUtil.kt:
New getDisplayTrackNumber() utility for consistent track number display across app, including fallback to positional numbering.

SimpleSongAdapter.kt:
Uses new utility for track numbers, replacing "" with always-valid numbers.

Screenshots of fix:
Before:
![telegram-cloud-photo-size-5-6282535258896927826-y](https://github.com/user-attachments/assets/b756dcea-6fbe-4df3-8abf-e082fea9e97a)

After:
![telegram-cloud-photo-size-5-6282535258896927825-y](https://github.com/user-attachments/assets/c126371f-5ab6-4157-a38c-97678c64077c)

gurnoorpannu committed:
feat: Fix individual song covers and track numbering display

🎵 Individual Song Cover Art Support
- Prioritize individual song embedded artwork over album-based covers
- Implement robust fallback strategy with comprehensive error handling
- Support both MediaMetadataRetriever and JAudioTagger for metadata extraction
- Add external cover art file detection (cover.jpg, album.jpg, etc.)
- Prevent cover art system crashes with graceful exception handling

🔢 Enhanced Track Numbering Logic
- Fix track numbers showing as '-' instead of proper numbering
- Implement intelligent track number display with multiple fallback strategies
- Support iTunes-style track numbering (e.g., 1002 → track 2)
- Handle single-song albums (always show '1')
- Use position-based numbering for missing metadata

🏗️ Professional Code Quality Improvements
- Comprehensive KDoc/JavaDoc documentation for all public APIs
- Input validation with descriptive error messages
- Proper resource management and cleanup
- Thread-safe operations with exception boundaries
- Performance optimizations with file size validation
- Clean architecture patterns and separation of concerns

📁 Files Modified:
- RetroGlideExtension.kt: Enhanced song model loading logic
- AudioFileCoverFetcher.kt: Robust artwork extraction with fallbacks
- AudioFileCoverUtils.java: Comprehensive metadata and external file support
- MusicUtil.kt: Centralized track number display logic
- SimpleSongAdapter.kt: Improved track numbering in album views
- AudioFileCover.kt: Enhanced model with validation
- AudioFileCoverLoader.kt: Professional Glide integration

✅ Expected Results:
- Individual songs now display unique cover art when available
- Track numbers show as 1, 2, 3, etc. instead of dashes
- Graceful fallback to album art when individual covers unavailable
- No more cover art system crashes from corrupted files
- Consistent track numbering across all album views

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/glide/RetroGlideExtension.kt:
```diff
@@ -65,8 +65,31 @@ object RetroGlideExtension {
+    /**
+     * Gets the appropriate model for loading song artwork.
+     * 
+     * This method prioritizes individual song embedded artwork over album-based covers
```
It does not? This method just gets the song cover, doesn't it?
Please only add useful comments, and not just AI generated stuff

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/glide/audiocover/AudioFileCoverUtils.java:
```diff
@@ -28,37 +28,163 @@
+      
+    } catch (ReadOnlyFileException | InvalidAudioFrameException | 
+             TagException | IOException | CannotReadException e) {
+      // Tag extraction failed, will try external files
```
Why are there two empty catch blocks?

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/glide/audiocover/AudioFileCoverFetcher.kt:
```diff
@@ -22,23 +22,108 @@ import java.io.FileNotFoundException
+    private fun handleFallbackOrFail(
+        callback: DataFetcher.DataCallback<in InputStream>,
+        originalException: Exception,
+        context: String
```
context is not used

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/glide/audiocover/AudioFileCoverFetcher.kt:
```diff
@@ -22,23 +22,108 @@ import java.io.FileNotFoundException
+                NoSuchElementException("No cover art found for file: ${model.filePath}")
+            )
+            
+        } catch (e: SecurityException) {
```
All catch blocks do the same, it can be one

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/util/MusicUtil.kt:
```diff
@@ -142,6 +142,44 @@ object MusicUtil : KoinComponent {
+            fixedTrackNumber > 0 -> fixedTrackNumber.toString()
+            
+            // Single song context - always show "1"
+            totalSongs == 1 -> "1"
```
Can't we just use the position in this case?

gurnoorpannu replied:
it was in the code already and i thought not to change it , also the numbering was working fine with only 1 song in album but if the song number increase more than one then there was the issue

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/util/MusicUtil.kt:
```diff
@@ -142,6 +142,44 @@ object MusicUtil : KoinComponent {
+     */
+    @JvmStatic
+    fun getDisplayTrackNumber(song: Song, position: Int, totalSongs: Int): String {
+        require(position >= 0) { "Position must be non-negative, got: $position" }
```
Why these checks, this should not fail, and if no track number can be determined, it can be "-" or sth

tomaThomas commented:
Please don't just let an AI generate comments, they are way too verbose, only add meaningful comments e.g. explaining complex code
Also, AI tends to put whatever is in the prompt in the comment, which often is not relevant when reading the code

gurnoorpannu commented:
> Please don't just let an AI generate comments, they are way too verbose, only add meaningful comments e.g. explaining complex code Also, AI tends to put whatever is in the prompt in the comment, which often is not relevant when reading the code

oh!okay  i thought it will be easier for the moderators to understand, but will make sure from now on to limit to relevant comments

gurnoorpannu committed:
requested changes for closing issue #1823

gurnoorpannu committed:
further changes fir #1843

gurnoorpannu commented:
@tomaThomas can you review this now?

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/glide/RetroGlideExtension.kt:
```diff
@@ -66,7 +66,15 @@ object RetroGlideExtension {
+        }
+    }
+
+    fun getAlbumModel(song: Song): Any {
```
Unused

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/glide/audiocover/AudioFileCoverFetcher.kt:
```diff
@@ -22,39 +22,79 @@ import java.io.FileNotFoundException
+    
     private var stream: InputStream? = null
+    
+    companion object {
```
Unused

tomaThomas commented on app/src/main/java/code/name/monkey/retromusic/glide/RetroGlideExtension.kt:
```diff
@@ -66,7 +66,15 @@ object RetroGlideExtension {
     }
 
     fun getSongModel(song: Song): Any {
-        return getSongModel(song, PreferenceUtil.isIgnoreMediaStoreArtwork)
```
So this setting does not do anything anymore? What does this PR change, isn't it basically ignoring MediaStore artwork all of the time?

gurnoorpannu committed:
Fix individual song covers implementation

- Respect PreferenceUtil.isIgnoreMediaStoreArtwork setting
- Add safety check for blank file paths to prevent crashes
- Remove unused code (getAlbumModel function, companion object)
- Clean up unused imports and consolidate exception handling
- Maintain backward compatibility with existing MediaStore behavior
